src/Controller/DownloadController.php

```diff
@@ -1,45 +1,99 @@
+
+            if ($winstable && $winpre && ($winpre[0]->getDate() < $winstable[0]->getDate()))
                 $winpre = null;
-            if ($osxstable && $osxpre && ($osxpre[0]['created_at'] < $osxstable[0]['created_at']))
+            if ($osxstable && $osxpre && ($osxpre[0]->getDate() < $osxstable[0]->getDate()))
                 $osxpre = null;
-            if ($linstable && $linpre && ($linpre[0]['created_at'] < $linstable[0]['created_at']))
+            if ($linstable && $linpre && ($linpre[0]->getDate() < $linstable[0]->getDate()))
                 $linpre = null;
-    
+
+            $assets = $artifacts->getForBranch('master');
+            $winnightly = array_values(array_filter($assets, self::assetsForPlatform(Platform::Windows)));
+            $osxnightly = array_values(array_filter($assets, self::assetsForPlatform(Platform::MacOS)));
+            $linnightly = array_values(array_filter($assets, self::assetsForPlatform(Platform::Linux)));
+
             $vars = [
                 'winstable' => $winstable,
                 'winpre' => $winpre,
+                'winnightly' => $winnightly,
                 'osxstable' => $osxstable,
                 'osxpre' => $osxpre,
+                'osxnightly' => $osxnightly,
                 'linstable' => $linstable,
-                'linpre' => $linpre
+                'linpre' => $linpre,
+                'linnightly' => $linnightly
+            ];
+            return $this->render('download/index.twig', $vars);
+        } catch (\Exception $e) {
+            error_log($e);
+            return $this->render('download/error.twig');
+        }
+    }
+
+    public function pull_request(string $id, Artifacts $artifacts): Response
+    {
+        try {
+            $assets = $artifacts->getForPullRequest($id);
+            $winartifacts = array_values(array_filter($assets, self::assetsForPlatform(Platform::Windows)));
+            $osxartifacts = array_values(array_filter($assets, self::assetsForPlatform(Platform::MacOS)));
+            $linartifacts = array_values(array_filter($assets, self::assetsForPlatform(Platform::Linux)));
+
+            $vars = [
+                'id' => $id,
+                'winartifacts' => $winartifacts,
+                'osxartifacts' => $osxartifacts,
+                'linartifacts' => $linartifacts
             ];
-        } catch (Exception $e) {
+            return $this->render('download/pull-request.twig', $vars);
+        } catch (\Exception $e) {
+            if ($e->getCode() === 404) {
+                throw $this->createNotFoundException(previous: $e);
+            }
             error_log($e);
             return $this->render('download/error.twig');
         }
-    
-        return $this->render('download/index.twig', $vars);
+    }
+
+    public function artifact(string $id, Artifacts $artifacts): Response
+    {
+        try {
+            return $this->redirect($artifacts->getArtifactDownloadUrl($id));
```

Comment:
This line occasionally throws the error:
```
{"count":11,"value":"Uri expired"}
```

This may be a caching issue, but it's certainly one we aren't accounting for in code.

Reply:
Opened new issue #362.